### ISSUE TITLE
Update with doc_as_upsert + additional fields.

### ISSUE THREAD

**Issue by d-ph** [elastic/elasticsearch#5889]
Hello,

how about adding a feature to update api, that allows you to perform an update with upsert defined as "the same doc plus additional fields".

Currently this can be done with this:

```
{
  doc: {
    field1: "Lorem ipsum"
  },
  upsert: {
    field1: "Lorem ipsum",
    additional_field1: 10
  }
}
```

This solution suffers from the same issue as one described here: #3153 (namely: sending potentially large data amount over the wire twice). Maybe, when in update body there are both `doc_as_upsert` and `upsert`, Elasticsearch should merge the sent `doc` with the hashmap in `upsert`? That's what I actually tried to do, but ES in this case favors `doc_as_upsert` over `upsert` content.

So, basically something like this would be nice:

```
{
  doc: {
    field1: "Lorem ipsum"
  },
  doc_as_upsert: true,
  upsert: {
    additional_field1: 10
  }
}
```

Cheers,
d

**Comment by dakrone** [comment]
This sounds interesting, however we discussed it and I think it would be better as part of revisiting the update API in general for API cleanups/enhancements. Marking this as "adopt me" right now and we will fold it into enhancing the update API in the future as well.

**Comment by clintongormley** [comment]
This request hasn't gained any traction, so I'm going to close it for now
